### PR TITLE
Add binary_response_with_content_type

### LINKED ISSUE #10: [Suggestion] Enable "binary_response" to Set a Content Type
https://github.com/rory/rust-cgi/blob/269b6680b84bc2645e89ec144f31bbc8b7a4bb2a/src/lib.rs#L218-L228

I suggest enabling `binary_response` to get content type as a parameter and to set it as the response.

### PR DESCRIPTION
**Pull request by loloicci**
fix #10 to add another function binary_response_with_context_type

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -223,10 +223,23 @@ pub fn binary_response<T>(status_code: T, body: Vec<u8>) -> Response
     http::response::Builder::new()
         .status(status_code)
         .header(http::header::CONTENT_LENGTH, format!("{}", body.len()).as_str())
+        .header(http::header::CONTENT_TYPE, "application/octet-stream")
         .body(body)
         .unwrap()
 }
 
+/// Sends  `blob` with that status code.
+pub fn binary_response_with_context_type<T>(status_code: T, body: Vec<u8>, context_type: &str) -> Response
```
Comment (rory) [inline src/lib.rs:232]: Can you make the `body` be the last argument, that seems more natural

Comment (rory) [inline src/lib.rs:232]: The HTTP header is called `Content-Type`, you're using `conte**x**t type` here. Can you use `content_type`?

Comment (rory) [inline src/lib.rs:232]: You're passing in a raw `&str` as the content type. [`.header(…)`](https://docs.rs/http/0.2.1/http/response/struct.Builder.html#method.header) can take a `TryFrom` for `HeaderValue`, which supports `&str` and other things too. Can you change the argument type to that?

File: src/lib.rs [hunk 2]
```diff
@@ -223,10 +223,23 @@ pub fn binary_response<T>(status_code: T, body: Vec<u8>) -> Response
     http::response::Builder::new()
         .status(status_code)
         .header(http::header::CONTENT_LENGTH, format!("{}", body.len()).as_str())
+        .header(http::header::CONTENT_TYPE, "application/octet-stream")
```
Comment (rory) [inline src/lib.rs:226]: Is this always the case? 🤔

**Comment by rory** [comment]
This is a good idea. Can you also add some unittests & add a line to the CHANGELOG file too.

File: src/lib.rs (hunk 1 above)
Comment (loloicci) [reply src/lib.rs:232]: Do you mean as follows?

```
/// Sends  `blob` with that status code and content type.
pub fn binary_response_with_content_type<T, V>(status_code: T, content_type: V, body: Vec<u8>) -> Response
    where
        http::StatusCode: TryFrom<T>,
        <http::StatusCode as TryFrom<T>>::Error: Into<http::Error>,
        http::HeaderValue: TryFrom<V>,
        <http::HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
{
    http::response::Builder::new()
        .status(status_code)
        .header(http::header::CONTENT_LENGTH, format!("{}", body.len()).as_str())
        .header(http::header::CONTENT_TYPE, content_type)
        .body(body)
        .unwrap()
}
```

This code causes a compile error.

```
240 |         .header(http::header::CONTENT_LENGTH, format!("{}", body.len()).as_str())
    |                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter `V`, found `&str`
```

If it is what you want to do, do you have any idea to solve this?

File: src/lib.rs (hunk 2 above)
Comment (loloicci) [reply src/lib.rs:226]: >  Any HTTP/1.1 message containing an entity-body SHOULD include a
   Content-Type header field defining the media type of that body. If
   and only if the media type is not given by a Content-Type field, the
   recipient MAY attempt to guess the media type via inspection of its
   content and/or the name extension(s) of the URI used to identify the
   resource. If the media type remains unknown, the recipient SHOULD
   treat it as type "application/octet-stream".

(quoted from RFC 2616, 7.2.1 Type)

I think, it should be "application/octet-strem" if we implement `binary_response_with_content_type`.

Comment (loloicci) [reply src/lib.rs:226]: thanks, I will delete it in the next commit.

File: src/lib.rs (hunk 1 above)
Comment (rory) [reply src/lib.rs:232]: oh yeah there's somethng funny about type inference, i remember that

**Comment by rory** [comment]
merged in

**Pull request closed without merging**